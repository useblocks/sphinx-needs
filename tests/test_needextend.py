import json
import sys
from pathlib import Path

import pytest
from sphinx.application import Sphinx
from syrupy.filters import props


@pytest.mark.parametrize("test_app", [{"buildername": "html", "srcdir": "doc_test/doc_needextend"}], indirect=True)
def test_doc_needextend_html(test_app: Sphinx, snapshot):
    app = test_app
    app.build()

    needs_data = json.loads(Path(app.outdir, "needs.json").read_text())
    assert needs_data == snapshot(exclude=props("created"))

    index_html = Path(app.outdir, "index.html").read_text()
    assert "extend_test_003" in index_html

    assert (
        '<div class="line">links outgoing: <span class="links"><span><a class="reference internal" href="#extend_'
        'test_004" title="extend_test_003">extend_test_004</a></span></span></div>' in index_html
    )

    assert (
        '<div class="line">links outgoing: <span class="links"><span><a class="reference internal" href="#extend_'
        'test_003" title="extend_test_006">extend_test_003</a>, <a class="reference internal" href="#extend_'
        'test_004" title="extend_test_006">extend_test_004</a></span></span></div>' in index_html
    )

    page_1__html = Path(app.outdir, "page_1.html").read_text()
    assert (
        '<span class="needs_data_container"><span class="needs_data">tag_1</span><span class="needs_spacer">, '
        '</span><span class="needs_data">new_tag</span><span class="needs_spacer">, '
        '</span><span class="needs_data">another_tag</span></span>' in page_1__html
    )


@pytest.mark.parametrize(
    "test_app", [{"buildername": "html", "srcdir": "doc_test/doc_needextend_strict"}], indirect=True
)
def test_doc_needextend_strict(test_app):
    import os
    import subprocess

    app = test_app

    srcdir = Path(app.srcdir)
    out_dir = os.path.join(srcdir, "_build")

    out = subprocess.run(["sphinx-build", "-b", "html", srcdir, out_dir], capture_output=True)

    # Strict option is set to false on needextend. Log info-level message
    assert "Provided id strict_disable_extend_test for needextend does not exist." in out.stdout.decode("utf-8")
    # Strict option is set to true on needextend. Raise Exception
    if sys.platform == "win32":
        assert (
            "Sphinx error:\r\nProvided id strict_enable_extend_test for needextend does not exist."
            in out.stderr.decode("utf-8")
        )
    else:
        assert (
            "Sphinx error:\nProvided id strict_enable_extend_test for needextend does not exist."
            in out.stderr.decode("utf-8")
        )


@pytest.mark.parametrize(
    "test_app", [{"buildername": "html", "srcdir": "doc_test/doc_needextend_dynamic"}], indirect=True
)
def test_doc_needextend_dynamic(test_app, snapshot):
    app = test_app
    app.build()

    # for some reason this intermittently creates incorrect warnings about overriding visitors
    # assert app._warning.getvalue() == ""

    needs_data = json.loads(Path(app.outdir, "needs.json").read_text())
    assert needs_data == snapshot(exclude=props("created"))
