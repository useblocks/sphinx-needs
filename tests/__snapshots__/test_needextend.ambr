# serializer version: 1
# name: test_doc_needextend_html[test_app0]
  dict({
    'current_version': '',
    'project': 'Python',
    'versions': dict({
      '': dict({
        'filters': dict({
        }),
        'filters_amount': 0,
        'needs': dict({
          'extend_test_003': dict({
            'arch': dict({
            }),
            'avatar': '',
            'closed_at': '',
            'completion': '',
            'constraints': list([
            ]),
            'constraints_passed': True,
            'constraints_results': dict({
            }),
            'content_id': 'extend_test_003',
            'created_at': '',
            'delete': None,
            'description': '''
              Had no outgoing links.
              Got an outgoing link ``extend_test_004``.
            ''',
            'docname': 'index',
            'doctype': '.rst',
            'duration': '',
            'external_css': 'external_link',
            'external_url': None,
            'full_title': 'needextend Example 3',
            'has_dead_links': '',
            'has_forbidden_dead_links': '',
            'hidden': '',
            'id': 'extend_test_003',
            'id_prefix': '',
            'is_external': False,
            'is_modified': True,
            'is_need': True,
            'is_part': False,
            'jinja_content': None,
            'layout': '',
            'links': list([
              'extend_test_004',
            ]),
            'max_amount': '',
            'max_content_lines': '',
            'modifications': 1,
            'params': '',
            'parent_need': '',
            'parent_needs': list([
            ]),
            'parent_needs_back': list([
            ]),
            'parts': dict({
            }),
            'post_template': None,
            'pre_template': None,
            'prefix': '',
            'query': '',
            'section_name': 'Need extend',
            'sections': list([
              'Need extend',
            ]),
            'service': '',
            'signature': '',
            'specific': '',
            'status': None,
            'style': None,
            'tags': list([
            ]),
            'target_id': 'extend_test_003',
            'template': None,
            'title': 'needextend Example 3',
            'type': 'story',
            'type_name': 'User Story',
            'updated_at': '',
            'url': '',
            'url_postfix': '',
            'user': '',
          }),
          'extend_test_004': dict({
            'arch': dict({
            }),
            'avatar': '',
            'closed_at': '',
            'completion': '',
            'constraints': list([
            ]),
            'constraints_passed': True,
            'constraints_results': dict({
            }),
            'content_id': 'extend_test_004',
            'created_at': '',
            'delete': None,
            'description': '''
              Had no links.
              Got an incoming links ``extend_test_003`` and ``extend_test_006``.
            ''',
            'docname': 'index',
            'doctype': '.rst',
            'duration': '',
            'external_css': 'external_link',
            'external_url': None,
            'full_title': 'needextend Example 4',
            'has_dead_links': '',
            'has_forbidden_dead_links': '',
            'hidden': '',
            'id': 'extend_test_004',
            'id_prefix': '',
            'is_external': False,
            'is_modified': False,
            'is_need': True,
            'is_part': False,
            'jinja_content': None,
            'layout': '',
            'links': list([
            ]),
            'max_amount': '',
            'max_content_lines': '',
            'modifications': 0,
            'params': '',
            'parent_need': '',
            'parent_needs': list([
            ]),
            'parent_needs_back': list([
            ]),
            'parts': dict({
            }),
            'post_template': None,
            'pre_template': None,
            'prefix': '',
            'query': '',
            'section_name': 'Need extend',
            'sections': list([
              'Need extend',
            ]),
            'service': '',
            'signature': '',
            'specific': '',
            'status': None,
            'style': None,
            'tags': list([
            ]),
            'target_id': 'extend_test_004',
            'template': None,
            'title': 'needextend Example 4',
            'type': 'story',
            'type_name': 'User Story',
            'updated_at': '',
            'url': '',
            'url_postfix': '',
            'user': '',
          }),
          'extend_test_005': dict({
            'arch': dict({
            }),
            'avatar': '',
            'closed_at': '',
            'completion': '',
            'constraints': list([
            ]),
            'constraints_passed': True,
            'constraints_results': dict({
            }),
            'content_id': 'extend_test_005',
            'created_at': '',
            'delete': None,
            'description': '''
              Had the two links: ``extend_test_003`` and ``extend_test_004``.
              Both got deleted.
            ''',
            'docname': 'index',
            'doctype': '.rst',
            'duration': '',
            'external_css': 'external_link',
            'external_url': None,
            'full_title': 'needextend Example 5',
            'has_dead_links': '',
            'has_forbidden_dead_links': '',
            'hidden': '',
            'id': 'extend_test_005',
            'id_prefix': '',
            'is_external': False,
            'is_modified': True,
            'is_need': True,
            'is_part': False,
            'jinja_content': None,
            'layout': '',
            'links': list([
            ]),
            'max_amount': '',
            'max_content_lines': '',
            'modifications': 1,
            'params': '',
            'parent_need': '',
            'parent_needs': list([
            ]),
            'parent_needs_back': list([
            ]),
            'parts': dict({
            }),
            'post_template': None,
            'pre_template': None,
            'prefix': '',
            'query': '',
            'section_name': 'Need extend',
            'sections': list([
              'Need extend',
            ]),
            'service': '',
            'signature': '',
            'specific': '',
            'status': None,
            'style': None,
            'tags': list([
            ]),
            'target_id': 'extend_test_005',
            'template': None,
            'title': 'needextend Example 5',
            'type': 'story',
            'type_name': 'User Story',
            'updated_at': '',
            'url': '',
            'url_postfix': '',
            'user': '',
          }),
          'extend_test_006': dict({
            'arch': dict({
            }),
            'avatar': '',
            'closed_at': '',
            'completion': '',
            'constraints': list([
            ]),
            'constraints_passed': True,
            'constraints_results': dict({
            }),
            'content_id': 'extend_test_006',
            'created_at': '',
            'delete': None,
            'description': 'Had the link ``extend_test_003``, got another one ``extend_test_004``.',
            'docname': 'index',
            'doctype': '.rst',
            'duration': '',
            'external_css': 'external_link',
            'external_url': None,
            'full_title': 'needextend Example 6',
            'has_dead_links': '',
            'has_forbidden_dead_links': '',
            'hidden': '',
            'id': 'extend_test_006',
            'id_prefix': '',
            'is_external': False,
            'is_modified': True,
            'is_need': True,
            'is_part': False,
            'jinja_content': None,
            'layout': '',
            'links': list([
              'extend_test_003',
              'extend_test_004',
            ]),
            'max_amount': '',
            'max_content_lines': '',
            'modifications': 2,
            'params': '',
            'parent_need': '',
            'parent_needs': list([
            ]),
            'parent_needs_back': list([
            ]),
            'parts': dict({
            }),
            'post_template': None,
            'pre_template': None,
            'prefix': '',
            'query': '',
            'section_name': 'Need extend',
            'sections': list([
              'Need extend',
            ]),
            'service': '',
            'signature': '',
            'specific': '',
            'status': None,
            'style': None,
            'tags': list([
            ]),
            'target_id': 'extend_test_006',
            'template': None,
            'title': 'needextend Example 6',
            'type': 'story',
            'type_name': 'User Story',
            'updated_at': '',
            'url': '',
            'url_postfix': '',
            'user': '',
          }),
          'extend_test_007': dict({
            'arch': dict({
            }),
            'avatar': '',
            'closed_at': '',
            'completion': '',
            'constraints': list([
            ]),
            'constraints_passed': True,
            'constraints_results': dict({
            }),
            'content_id': 'extend_test_007',
            'created_at': '',
            'delete': None,
            'description': '',
            'docname': 'index',
            'doctype': '.rst',
            'duration': '',
            'external_css': 'external_link',
            'external_url': None,
            'full_title': 'needextend Example 7',
            'has_dead_links': '',
            'has_forbidden_dead_links': '',
            'hidden': '',
            'id': 'extend_test_007',
            'id_prefix': '',
            'is_external': False,
            'is_modified': True,
            'is_need': True,
            'is_part': False,
            'jinja_content': None,
            'layout': '',
            'links': list([
              'extend_test_003',
              'extend_test_004',
              'extend_test_005',
            ]),
            'max_amount': '',
            'max_content_lines': '',
            'modifications': 1,
            'params': '',
            'parent_need': '',
            'parent_needs': list([
            ]),
            'parent_needs_back': list([
            ]),
            'parts': dict({
            }),
            'post_template': None,
            'pre_template': None,
            'prefix': '',
            'query': '',
            'section_name': 'Need extend',
            'sections': list([
              'Need extend',
            ]),
            'service': '',
            'signature': '',
            'specific': '',
            'status': None,
            'style': None,
            'tags': list([
            ]),
            'target_id': 'extend_test_007',
            'template': None,
            'title': 'needextend Example 7',
            'type': 'story',
            'type_name': 'User Story',
            'updated_at': '',
            'url': '',
            'url_postfix': '',
            'user': '',
          }),
          'extend_test_page': dict({
            'arch': dict({
            }),
            'avatar': '',
            'closed_at': '',
            'completion': '',
            'constraints': list([
            ]),
            'constraints_passed': True,
            'constraints_results': dict({
            }),
            'content_id': 'extend_test_page',
            'created_at': '',
            'delete': None,
            'description': '',
            'docname': 'page_1',
            'doctype': '.rst',
            'duration': '',
            'external_css': 'external_link',
            'external_url': None,
            'full_title': 'needextend different pages test',
            'has_dead_links': '',
            'has_forbidden_dead_links': '',
            'hidden': '',
            'id': 'extend_test_page',
            'id_prefix': '',
            'is_external': False,
            'is_modified': True,
            'is_need': True,
            'is_part': False,
            'jinja_content': None,
            'layout': '',
            'links': list([
            ]),
            'max_amount': '',
            'max_content_lines': '',
            'modifications': 2,
            'params': '',
            'parent_need': '',
            'parent_needs': list([
            ]),
            'parent_needs_back': list([
            ]),
            'parts': dict({
            }),
            'post_template': None,
            'pre_template': None,
            'prefix': '',
            'query': '',
            'section_name': 'need objects',
            'sections': list([
              'need objects',
            ]),
            'service': '',
            'signature': '',
            'specific': '',
            'status': 'closed',
            'style': None,
            'tags': list([
              'tag_1',
              'new_tag',
              'another_tag',
            ]),
            'target_id': 'extend_test_page',
            'template': None,
            'title': 'needextend different pages test',
            'type': 'story',
            'type_name': 'User Story',
            'updated_at': '',
            'url': '',
            'url_postfix': '',
            'user': '',
          }),
        }),
        'needs_amount': 6,
      }),
    }),
  })
# ---
